### PR TITLE
Migration to the Refactored LArG4

### PR DESCRIPTION
**Pull request by marcodeltutto**
This PR contains all the necessary fcls and code to migrate to the new LArG4. 
Changes are presented on docdb [22944](https://sbn-docdb.fnal.gov/cgi-bin/sso/ShowDocument?docid=22944).

Note that with this PR, the `standard_g4_sbnd.fcl` will run the new LArG4. To run the legacy larg4, fcl `legacy_g4_sbnd.fcl` should be used. All the `g4_*.fcl` files run the new LArG4. To run the legacy LArG4, replace `g4_*.fcl` with `legacy_g4_*.fcl`.
All the legacy g4 fcl files are in `JobConfigurations/standard/g4/legacy/*`

A new folder called `LArG4` is added. This contains all the g4-related fcl configuration, as well as the physics list to be used in `LArG4/PhysicsLists/`. In this `LArG4` folder:
- `larg4_sbnd.fcl`: Configuration of the LArG4 module
- `larg4_services_sbnd.fcl`: Configurations needed for the new LArG4 
- `ionandscint_sbnd.fcl`: Configuration for the ionization and scintillation part
- `simdrift_sbnd.fcl`: Configuration for the electron drifting module
- `mcreco_sbnd.fcl`: Configuration for MC reconstruction
- `PhysicsListLoader`: a dummy module to load the physics list, needed before running LArG4

To run with the optical hybrid mode, fcl file `g4_ophybrid_sbnd.fcl` can be used in place of `standard_g4_sbnd.fcl`.

### Open Items

- [x] Use **correlated** light model? -> Yes, as PR #127 was merged.
- [x] Use **optical hybrid** mode? -> Added to default g4 workflow.
- [x] ~~**CRT** part. -> Waiting for PR to be opened.~~

### FHICL files
Suggested reviewer: @jzennamo 
Standard fcls now run the new larg4. All the legacy fcls are in `sbndcode/JobConfigurations/standard/g4/legacy/`, and similar for `detsim`.

### Optical
Suggested reviewer: @ggamezdiego  
- `sbndcode/JobConfigurations/standard/standard_g4_sbnd.fcl`: Runs in optical hybrid mode by default.
- `sbndcode/LArSoftConfigurations/photpropservices_sbnd.fcl`: Adds optical library for light outside the active volume.
- `sbndcode/LArSoftConfigurations/PDFastSim_sbnd.fcl`: Configuration for semi-analytical and optical library. 
- `sbndcode/LArG4/larg4_services_sbnd.fcl`: Specifies the LArG4 detector to use for the optical hybrid mode.
- `sbndcode/JobConfigurations/standard/g4/g4_noophybrid_sbnd.fcl`: Same as standard_g4_sbnd.fcl, but without the optical hybrid part.

### CRT
CRT will be included next. Right now, this PR excludes the CRT from the detsim stage. Issue #143  has been opened to track this.

### Known Issues
- All the MCParticles are saved, not only the ones in the cryostat and CRT. LArSoft issue [26106](https://cdcvs.fnal.gov/redmine/issues/26106). For now, I produced GENIE+cosmic files and verified that the file size, although larger with the new LArG4, is still manageable.
- There are many warning messages from `SimDriftElectrons` of type `[...] unable to drift electrons from point [...]` when electrons are drifted from points outside the active volume. LArSoft issue [26107](https://cdcvs.fnal.gov/redmine/issues/26107).
- The `g4_simphotontime_filter.fcl` will have to be reviewed to make sure the in-time and out-time paths are both running correctly. Issue #147 was opened to keep track of this issue.

### CODE REVIEW

**Comment by ggamezdiego** [comment]
All the hybrid-optical .fcl files look ok to me. Only one thing. I have not seen the corrections from the LArQL model set true (LArG4Parameters.UseModLarqlRecomb). Did we decide not to include them? It represents a 20% difference for the light generated outside the TPC (inside the difference is negligible). Or maybe this is set true somewhere else?

**Comment by wforeman** [comment]
@ggamezdiego The LArQL model will alter the simulated charge deposition in addition to the light (since it modifies the recombination directly)... was the plan to only enable it outside the TPC?

**Comment by ggamezdiego** [comment]
@wforeman, my understanding is that it is only sensitive for small values of the EField. We verified that inside the TPC (EF=500V/cm) the effect on the light-yield is really marginal, and this should also be the case for the charge. Only outside the TPC (so irrelevant for Q), where we have set our EF=0, it shows an effect in L (~20% reduction). So Q shouldn't be affected in any case

**Comment by marcodeltutto** [comment]
Thanks @ggamezdiego! I don't think we came to a conclusion about the LArQL, this is the only reason why I didn't add it to this pull request. I can add it if people agree, or we can add it in the Oct production if we don't want to have all the changes together.

**Comment by jzennamo** [comment]
My understanding was that the plan was to use Will's Q+L model (https://github.com/SBNSoftware/sbndcode/pull/127) but not the LArQL for the coming production. I think we can bring up if it makes sense to use LArQL for the October production, does that sound reasonable? (Could someone make an issue for Oct to track this?)

File: sbndcode/JobConfigurations/standard/standard_detsim_sbnd.fcl [hunk 1]
```diff
@@ -78,14 +65,15 @@ physics:
     opdaq:     @local::sbnd_opdetdigitizer
   }
 
-  #define the producer and filter modules for this path, order matters, 
-  simulate:  [ rns, daq, crt, opdaq]
+  #define the producer and filter modules for this path, order matters,
+  # simulate:  [ rns,  daq, crt, opdaq]
```
Comment (jzennamo) [inline sbndcode/JobConfigurations/standard/standard_detsim_sbnd.fcl:69]: Do we have an issue open for the removal of the CRT from the detsim?

**Review by jzennamo — APPROVED** [review]
I just had two little questions, otherwise I approve this!

File: sbndcode/Geometry/gdml/sbnd_v02_00_base.gdml [hunk 2]
```diff
@@ -4655,6 +4661,9 @@
 		<volume name="volPMT">
 			<materialref ref="LAr"/>
 			<solidref ref="PMTVolume"/>
```
[the remark this answers is not in the document]
Comment (marcodeltutto) [reply sbndcode/Geometry/gdml/sbnd_v02_00_base.gdml:4663]: `volPMT ` is just the PMT "mother" volume. It contains sub-volumes `vol_PMT_Back` (steel), and `vol_PMT_in` (vacuum).

File: sbndcode/JobConfigurations/standard/standard_detsim_sbnd.fcl (hunk 1 above)
Comment (marcodeltutto) [reply sbndcode/JobConfigurations/standard/standard_detsim_sbnd.fcl:69]: We do! Issue #143.

File: sbndcode/Geometry/gdml/sbnd_v02_00_base.gdml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jzennamo) [reply sbndcode/Geometry/gdml/sbnd_v02_00_base.gdml:4663]: perfect!

File: sbndcode/JobConfigurations/standard/standard_detsim_sbnd.fcl (hunk 1 above)
Comment (jzennamo) [reply sbndcode/JobConfigurations/standard/standard_detsim_sbnd.fcl:69]: you're my hero!

**Comment by pgreen135** [comment]
trigger build

**Comment by pgreen135** [comment]
trigger build

**Comment by pgreen135** [comment]
@marcodeltutto can you have a look over the CI test results above, are these expected? there are two failures (the detsim stage for each test)

**Comment by wesketchum** [comment]
I'm worried these are not expected errors:

```
60: %MSG-s ArtException:  PostEndJob 16-Aug-2021 12:36:13 CDT ModuleEndJob
61: ---- EventProcessorFailure BEGIN
62:   EventProcessor: an exception occurred during current event processing
63:   ---- ScheduleExecutionFailure BEGIN
64:     Path: ProcessingStopped.
65:     ---- ProductNotFound BEGIN
66:       getView: Found no products matching all criteria
67:       Looking for sequence of type: sim::SimChannel
68:       Looking for module label: simdrift
69:       Looking for productInstanceName: 
70:       The above exception was thrown while processing module SimWireSBND/daq run: 1 subRun: 0 event: 1
71:     ---- ProductNotFound END
72:     Exception going through path simulate
73:   ---- ScheduleExecutionFailure END
74: ---- EventProcessorFailure END
75: %MSG
76: Art has completed and will exit with status 1.
```

Looks like a bad fhicl?

**Comment by jzennamo** [comment]
As far as I understand it the CI tests that are failing are doing so because they are trying to run DetSim over the reference files, but since G4 changed I don't think that this should work since there are data products missing. From my reading of these test outputs we should be OK. 

Once this is merged we should update the reference files immediately.

**Comment by wesketchum** [comment]
Issue #149 filed to capture this.

**Pull request merged**